exercises/building_suffix_arrays/descr.md

```diff
@@ -0,0 +1,543 @@
+*h*-*buckets*, where all of the suffixes inside each bucket are *h*-equal,
+that is, they agree on their *h* first characters.
+
+We assume that a Boolean array *b*, which contains beginning-of-bucket
+marks, is also given.
+
+### Aim
+
+The aim of this stage is to reach a situation where the array *a* is
+*2h*-sorted. This is a more precise preorder: indeed, two suffixes that are
+*2h*-equal must also be *h*-equal. This means that we wish to *refine* the
+current partition. In other words, we wish to *divide* each *h*-bucket into
+one or more *2h*-buckets.
+
+To do so, we must move some suffixes to a different place in the array *a*.
+These moves take place only *within* an *h*-bucket, never from an
+*h*-bucket to another.
+We must also update the array *b* so as to mark the beginning
+of each *2h*-bucket.
+
+It is worth noting that the suffixes in the interval *[n-h+1, n)* will not
+move during this stage. Indeed, these suffixes have fewer than *h* characters,
+so, as far as they are concerned, *h*-order and *2h*-order are the same thing.
+Because *a* is *h*-sorted, each of these suffixes has already reached its
+final position in the array, and inhabits a singleton *h*-bucket.
+
+### First Insight
+
+How can we implement this stage efficiently, in time *O(n)*?
+
+Manber and Myers' first insight is that the necessary comparisons *have
+already been performed* during previous stages! Indeed, suppose that two
+suffixes *d1* and *d2* lie in the same *h*-bucket. In order to determine in
+what order they should appear at the end of this stage, we need to
+*2h*-compare them, that is, to compare their *2h* first characters. However,
+we already know that their *h* first characters coincide. So, we need to
+compare the *next* run of *h* characters, between offsets *h* (included) and
+*2h* (excluded). In other words, we need to *h*-compare the suffixes *d1+h*
+and *d2+h*. (Let us assume in this informal explanation that *d1+h* and *d2+h*
+are both less than *n*, so they are well-defined suffixes.)
+This is something that we have already done during previous stages!
+Since the array *a* is *h*-sorted, all we need to do is find out which of the
+suffixes *d1+h* and *d2+h* appears first in the array *a*. With appropriate
+preprocessing (see following remark), this can be done in constant time.
+
+*Remark.* The permutation *a* can be inverted in time *O(n)*. Its inverse is
+an array that maps every suffix *d* to its position in the array *a*. Once
+this inverse permutation has been computed, we can find out in constant time
+where in the array *a* a suffix *d* is stored.
+
+To sum up, Manber and Myers' first insight is that two suffixes can be
+*2h*-compared in constant time.
+
+Thus, a naïve idea would be to implement stage *h* simply by sorting the array
+*a* using an off-the-shelf sorting algorithm, such as Heapsort or Mergesort.
+The total complexity of this stage would then be *O(n log n)*, which is not
```

Comment:
Is your grader able to detect if the student has submitted such an algorithm?

Reply:
No, the grader does not perform an explicit complexity check. The student's code is tested with relatively long strings (up to several hundred thousand characters) so its complexity must not be awful, or it will time out.

It's not clear how we would reliably measure the theoretical complexity of the code, as we do not have any obvious hooks. (Perhaps we could override the primitive array access operations and count them? But we would also need to prevent the student from calling library functions such as `List.sort` which perform array accesses without going through this wrapper.)